### DISCUSSION TITLE
Conditionally adding a value to an array based on previous trial data

### DISCUSSION THREAD

**Discussion by mcd710** [jspsych/jsPsych#2494] [Q&A] [answered]
Hi there! 

I've been working on a task for a few months now and I'm at the stage where I want to run it on Prolific, but I'm having trouble getting one last part to work. The task is a betting game, and I'm trying to award participants money based on their performance (and eventually display a screen at the end that says "you won x $"). There are four categories of images, and I'm trying to write a code that adds 0.05 to a pre-defined array of "money" based on whether they identified the image correctly. The names of the images are strings and I want to compare a subset of the strings (whether the string contains "ff", "fs", "ss", or "sf") to the response provided by the participant (either 0 or 1). I have three questions: (1) Since I have written and added ```countFunction()``` to my timeline, I keep getting the two errors ```"Trial level node is missing the "type" parameter"``` and ```"Uncaught TypeError: trial.type is not a constructor"``` – any ideas on how to fix this? (2) I tried to define my own function and put it in the timeline because I was not sure that there was a way within JsPsych to conditionally add data to an array based on the results of each trial. Is there a way I can use the JsPsych framework to accomplish this? (3) I'm not sure if I'm accessing the timeline data correctly when I, for example, write ```var image_data_ff_fs = jsPsych.data.getLastTimelineData().select('stimulus').subset(function(x){ return x = image_data.includes('ff' || 'fs')})```. Does this look correct? 

Any advice at all would be greatly appreciated. 

Here is a relevant section of my code:

```
    /* initialize jsPsych */
    var jsPsych = initJsPsych({
      on_finish: function() {
        jsPsych.data.displayData();
      }
    });
    
    /* create timeline */
    var timeline = [];
    
    /* create empty array for payment to be added to */
    var money = [];

    /* function to add the bonuses */
    // CALL THIS FUNCTION IN "STIMULUS" AT THE VERY END //
    function moneyFunction() {
        let sum = 0;
        for (let i = 0; i < money.length; i++) {
            sum += money[i];
        }
        console.log(sum);
    }

// removed a lot of code here //

    /* define trial stimuli array for timeline variables */
    var test_stimuli = [
        {stimulus: 'images/easy-411-ss.png'},
        {stimulus: 'images/easy-538-sf.png'},
        {stimulus: 'images/easy-vcom=4-126-fs.png'},
        {stimulus: 'images/hard-570-sf.png'},
        {stimulus: 'images/hard-600-ss.png'},
        {stimulus: 'images/hard-vcom=1-50-ff.png'},
        {stimulus: 'images/hard-vcom=1-95-fs.png'},
// there are a lot more images, I deleted most of them
    ];


    /* define test trials */
    var test = {
        type: jsPsychImageButtonResponse,
        stimulus: jsPsych.timelineVariable('stimulus'),
        choices: ['fall ◄', 'stand ►'],
        randomize_order: true,
        trial_duration: 5000,
        post_trial_gap: 500
    };

// removed a lot of code here ... includes test, if_node//

    /* get data from last trial and add to bonuses */
    function countFunction() {
        var response_data_0 = jsPsych.data.getLastTimelineData().select('response').subset(function(x){ return x == 0; });
        var response_data_1 = jsPsych.data.getLastTimelineData().select('response').subset(function(x){ return x == 1; });
        // is it actually called stimulus?
        var image_data_ff_fs = jsPsych.data.getLastTimelineData().select('stimulus').subset(function(x){ return x = image_data.includes('ff' || 'fs')});
        var image_data_ff_fs = jsPsych.data.getLastTimelineData().select('stimulus').subset(function(x){ return x = image_data.includes('ss' || 'sf')});
        if (response_data_0 && image_data_ff_fs) {
            money.push(0.05)
        }
        else if (response_data_1 && image_data_ff_fs) {
            money.push(0)
        }
        else if (response_data_1 && image_data_ss_sf) {
            money.push(0.05)
        }
        else if (response_data_0 && image_data_ss_sf) {
            money.push(0)
        }
    }


    /* test procedure */
    var test_procedure = {
        timeline: [test, if_node, countFunction()],
        timeline_variables: test_stimuli,
        randomize_order: true
    };
    timeline.push(test_procedure);
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @mcd710, for (1), you're getting the error because you can't put a function on the timeline - everything in the timeline array should be an object. To run your `countFunction` at that point in the timeline, you can either call the function (a) from inside the conditional function in your `if_node` (after your existing code but before the return statements), or (b) using the [`call-function` plugin](https://www.jspsych.org/7.1/plugins/call-function/). That should answer (2) as well.

For (3), I don't think what you have is quite right so I've suggested something else below. But I haven't tested this so it might not work. So to check, I've added statements to print out the values to the browser's JavaScript console.
```js
var image_data_ff_fs = jsPsych.data.getLastTimelineData().select('stimulus').subset(function(x){ 
  console.log('stimulus: ', x);
  console.log('includes ff or fs: ', x.includes('ff') || x.includes('fs'));
  return x.includes('ff') || x.includes('fs');
});
console.log('image data ff fs: ', image_data_ff_fs);
```
I hope this helps! Can you let us know if these solutions work for you?

**Reply by mcd710** [reply]
Thanks so much for your response! It makes a lot of sense. I included your edits and I tried using the call-function plugin, but I keep getting the error ```Uncaught ReferenceError: jsPsychCallFunction is not defined```. 

Is there something wrong with the way I'm implementing it? Could it be to do with the fact that my function has arguments? Sorry if there's an obvious error I'm missing!

```
    /* get data from last trial and add to bonuses */
    var countFunction = function() {
        var response_data_0 = jsPsych.data.getLastTimelineData().select('response').subset(function(x){ return x == 0; });
        var response_data_1 = jsPsych.data.getLastTimelineData().select('response').subset(function(x){ return x == 1; });
        // images that contain either ff or fs
        var image_data_ff_fs = jsPsych.data.getLastTimelineData().select('stimulus').subset(function(x){ 
            console.log('stimulus: ', x);
            console.log('includes ff or fs: ', x.includes('ff') || x.includes('fs'));
            return x.includes('ff') || x.includes('fs');
            console.log('image data ff fs: ', image_data_ff_fs);
            });
        // images that contain either ss or sf
        var image_data_ss_sf = jsPsych.data.getLastTimelineData().select('stimulus').subset(function(x){ 
            console.log('stimulus: ', x);
            console.log('includes ss or sf: ', x.includes('ss') || x.includes('sf'));
            return x.includes('ss') || x.includes('sf');
            console.log('image data ss sf: ', image_data_ss_sf);
            });
        if (response_data_0 && image_data_ff_fs) {
            money.push(0.05)
        }
        else if (response_data_1 && image_data_ff_fs) {
            money.push(0)
        }
        else if (response_data_1 && image_data_ss_sf) {
            money.push(0.05)
        }
        else if (response_data_0 && image_data_ss_sf) {
            money.push(0)
        }
    }

    var countingTrial = {
        type: jsPsychCallFunction,
        func: countFunction
    }


    /* test procedure */
    var test_procedure = {
        timeline: [test, if_node, countingTrial],
        timeline_variables: test_stimuli,
        randomize_order: true
    };
    timeline.push(test_procedure);

    console.log(money)
```

**Reply by becky-gilbert** [reply]
Hi @mcd710, that error suggests that the plugin file isn't loaded. Did you remember to add the call-function plugin to the scripts that you load in the `head` section of your HTML file? Or maybe there's a typo in that line?

**Reply by mcd710** [reply]
Oh my gosh, of course! Silly mistake. Thank you so much, everything works now!! :)
